src/main/java/edu/harvard/iq/dataverse/engine/command/impl/FinalizeDatasetPublicationCommand.java

```diff
@@ -237,10 +252,24 @@ private void updateParentDataversesSubjectsField(Dataset savedDataset, CommandCo
             if (dsf.getDatasetFieldType().getName().equals(DatasetFieldConstant.subject)) {
                 Dataverse dv = savedDataset.getOwner();
                 while (dv != null) {
-                    if (dv.getDataverseSubjects().addAll(dsf.getControlledVocabularyValues())) {
+                    boolean newSubjectsAdded = false;
+                    for (ControlledVocabularyValue cvv : dsf.getControlledVocabularyValues()) {
+                    
+                        if (!dv.getDataverseSubjects().contains(cvv)) {
+                            logger.fine("dv "+dv.getAlias()+" does not have subject "+cvv.getStrValue());
+                            newSubjectsAdded = true;
+                            dv.getDataverseSubjects().add(cvv);
+                        } else {
+                            logger.fine("dv "+dv.getAlias()+" already has subject "+cvv.getStrValue());
+                        }
+                    }
+                    if (newSubjectsAdded) {
+                        logger.fine("new dataverse subjects added - saving and reindexing");
                         Dataverse dvWithSubjectJustAdded = ctxt.em().merge(dv);
                         ctxt.em().flush();
                         ctxt.index().indexDataverse(dvWithSubjectJustAdded); // need to reindex to capture the new subjects
```

Comment:
I already opened a separate issue for this, but index should be moved to onSuccess.

Reply:
Makes sense.